controller.py

```diff
@@ -409,14 +409,70 @@ def register(self, do_get=HTTP.debuggable_get):
         if isinstance(integration_contact_email, ProblemDetail):
             return integration_contact_email
 
+        library = None
+        elevated_permissions = False
+        if shared_secret:
+            # Look up a library by the provided shared secret. This
+            # will let us handle the case where the library has
+            # changed URLs (auth_url does not match
+            # library.authentication_url) but the shared secret is the
+            # same.
+            library = get_one(self._db, Library, shared_secret=shared_secret)
+            if not library:
+                return AUTHENTICATION_FAILURE.detailed(
+                    _("Provided shared secret is invalid")
+                )
```

Comment:
We now check the shared secret immediately and return a problem detail if it's wrong; previously we were ignoring a bad shared secret unless the request did something that required elevated permissions.